src/sidebar/store/modules/toast-messages.js

```diff
@@ -0,0 +1,92 @@
+import * as util from '../util';
+
+/**
+ * A store module for managing a collection of toast messages. This module
+ * maintains state only; it's up to other layers to handle the management
+ * and interactions with these messages.
+ */
+
+function init() {
+  return {
+    messages: [],
+  };
+}
+
+const update = {
+  ADD_MESSAGE: function(state, action) {
+    return {
+      messages: state.messages.concat([{ ...action.message }]),
+    };
+  },
+
+  REMOVE_MESSAGE: function(state, action) {
+    const updatedMessages = state.messages.filter(
+      message => message.id !== action.id
+    );
+    return { messages: updatedMessages };
+  },
+
+  UPDATE_MESSAGE: function(state, action) {
+    const updatedMessages = state.messages.map(message => {
+      if (message.id && message.id === action.message.id) {
+        return { ...action.message };
+      }
+      return { ...message };
```

Comment:
I'm okay with this, but my preference is not to make a copy of the thing if its not changing only so its clear in code when we do make a copy, we are in fact making changes.

Reply:
You're right, I agree.